pkg/tasker/tasker.go

```diff
@@ -0,0 +1,169 @@
+	close            context.CancelFunc
+	logger           log.Logger
+	proxy            db.DataServerProxy
+	accounts         []*rpc.Account
+	contractInstance *contracts.ITellor
+	client           contracts.ETHClient
+	cfg              *config.Config
+	workSink         chan *mining.Work
+	Running          bool
+	done             chan bool
+	resubscribe      chan bool
+}
+
+func CreateTasker(ctx context.Context, logger log.Logger, cfg *config.Config, proxy db.DataServerProxy, client contracts.ETHClient, contract *contracts.ITellor, accounts []*rpc.Account) (*Tasker, chan *mining.Work) {
+	ctx, close := context.WithCancel(ctx)
+	tasker := &Tasker{
+		ctx:              ctx,
+		close:            close,
+		proxy:            proxy,
+		accounts:         accounts,
+		contractInstance: contract,
+		workSink:         make(chan *mining.Work, 1),
+		done:             make(chan bool),
+		resubscribe:      make(chan bool, 1),
+		logger:           log.With(logger, "component", ComponentName),
+		cfg:              cfg,
+		client:           client,
+	}
+	return tasker, tasker.workSink
+}
+
+func (mt *Tasker) getCurrentChallenge() (*tellor.ITellorNewChallenge, error) {
+	newVariables, err := mt.contractInstance.GetNewCurrentVariables(nil)
+	if err != nil {
+		level.Warn(mt.logger).Log("msg", "new current variables retrieval - contract might not be upgraded", "err", err)
+		return nil, err
+	}
+	return &tellor.ITellorNewChallenge{
+		CurrentChallenge: newVariables.Challenge,
+		Difficulty:       newVariables.Difficutly,
+		CurrentRequestId: newVariables.RequestIds,
+		TotalTips:        newVariables.Tip,
+	}, err
+}
+
+func (mt *Tasker) getNewChallengeChannel() (chan *tellor.ITellorNewChallenge, event.Subscription, error) {
+	sink := make(chan *tellor.ITellorNewChallenge)
+	var tellorFilterer *tellor.ITellorFilterer
+	tellorFilterer, err := tellor.NewITellorFilterer(mt.contractInstance.Address, mt.client)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting ITellorFilterer instance")
+	}
+	sub, err := tellorFilterer.WatchNewChallenge(&bind.WatchOpts{}, sink, nil)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting NewChallenge channel")
+	}
+	return sink, sub, nil
+}
+
+func (mt *Tasker) subscribeToNewChallenge() error {
+	var sink chan *tellor.ITellorNewChallenge
+	var err error
+	var sub event.Subscription
+	var failedSubscriptionCount int
+
+	// Initial subscription.
+	sink, sub, err = mt.getNewChallengeChannel()
+	if err != nil {
+		level.Error(mt.logger).Log("msg", "initial subscribing to NewChallenge events failed")
+		mt.resubscribe <- true
+	}
+	for {
+		select {
+		case <-mt.done:
+			if sub != nil {
+				sub.Unsubscribe()
+			}
+			level.Info(mt.logger).Log("msg", "unsubscribed to NewChallenge events")
+			return nil
+		case err := <-sub.Err():
+			if err != nil {
+				level.Error(mt.logger).Log(
+					"msg",
+					"new challenge subscription error",
+					"err", err)
+			}
+			mt.resubscribe <- true
```

Comment:
Why not get rid of the channel and just run a resubscribe func?

```suggestion
		resubscribe()
```

```
func ... resubscribe(){
	....
}
```

Reply:
Resolved in [hhio618/telliot#1](https://github.com/hhio618/telliot/pull/1/files).